src/sys/unix/uds.rs

```diff
@@ -0,0 +1,298 @@
+
+    syscall!(bind(socket, raw_addr, raw_addr_length))?;
+    Ok(unsafe { UnixDatagram::from_raw_fd(socket) })
+}
+
+// pub fn accept(listener: &UnixListener) -> io::Result<Option<(UnixStream, SocketAddr)>> {
+pub fn accept(listener: &UnixListener) -> io::Result<(UnixStream, SocketAddr)> {
+    let mut storage: libc::sockaddr_un = unsafe { mem::zeroed() };
+    storage.sun_family = libc::AF_UNIX as libc::sa_family_t;
+    let mut len = mem::size_of_val(&storage) as libc::socklen_t;
+    let raw_storage = &mut storage as *mut _ as *mut _;
+
+    #[cfg(not(any(
+        target_os = "ios",
+        target_os = "macos",
+        target_os = "netbsd",
+        target_os = "solaris"
+    )))]
+    let sock_addr = {
+        let flags = libc::SOCK_NONBLOCK | libc::SOCK_CLOEXEC;
+
+        match syscall!(accept4(listener.as_raw_fd(), raw_storage, &mut len, flags)) {
+            Ok(sa) => sa,
+            Err(e) => return Err(e),
+        }
+    };
+
+    #[cfg(any(
+        target_os = "ios",
+        target_os = "macos",
+        target_os = "netbsd",
+        target_os = "solaris"
+    ))]
+    let sock_addr = match syscall!(accept(listener.as_raw_fd(), raw_storage, &mut len)) {
+        Ok(sa) => sa,
+        Err(e) => return Err(e),
+    };
+
+    #[cfg(any(
+        target_os = "ios",
+        target_os = "macos",
+        target_os = "netbsd",
+        target_os = "solaris"
+    ))]
+    {
+        syscall!(fcntl(sock_addr, libc::F_SETFL, libc::O_NONBLOCK))?;
+        syscall!(fcntl(sock_addr, libc::F_SETFD, libc::FD_CLOEXEC))?;
+    }
+
+    Ok((
+        unsafe { UnixStream::from_raw_fd(sock_addr) },
+        SocketAddr { addr: storage, len },
+    ))
+}
+
+pub fn pair_datagram() -> io::Result<(UnixDatagram, UnixDatagram)> {
+    let fds = [0, 0];
+    let flags = libc::SOCK_DGRAM;
+
+    pair_descriptors(fds, flags)?;
+
+    Ok(unsafe {
+        (
+            UnixDatagram::from_raw_fd(fds[0]),
+            UnixDatagram::from_raw_fd(fds[1]),
+        )
+    })
+}
+
+pub fn unbound_datagram() -> io::Result<UnixDatagram> {
+    let socket = new_socket(libc::AF_UNIX, libc::SOCK_DGRAM)?;
+    Ok(unsafe { UnixDatagram::from_raw_fd(socket) })
+}
+
+pub fn bind_listener(path: &Path) -> io::Result<UnixListener> {
+    let socket = new_socket(libc::AF_UNIX, libc::SOCK_STREAM)?;
+    let (raw_addr, raw_addr_length) = socket_addr(path)?;
+    let raw_addr = &raw_addr as *const _ as *const _;
+
+    syscall!(bind(socket, raw_addr, raw_addr_length))
+        .and_then(|_| syscall!(listen(socket, 1024)))
+        .map_err(|err| {
+            // Close the socket if we hit an error, ignoring the error from
+            // closing since we can't pass back two errors.
+            let _ = unsafe { libc::close(socket) };
+            err
+        })
+        .map(|_| unsafe { UnixListener::from_raw_fd(socket) })
+}
+
+pub fn socket_addr(path: &Path) -> io::Result<(libc::sockaddr_un, libc::socklen_t)> {
+    unsafe {
```

Comment:
this is a pretty big wad of `unsafe`; can the unsafe block possibly be made more granular?

Reply:
So I reduced a lot of the unsafe here. It turns out that a `libc::sockaddr_un` zeroed out is a properly initialized value. I explained why in the comment I added, but the rest of the expressions that set the field values will not be UB. Let me know if that makes sense.

I also reduced the unsafe usage in `sun_path_offset` and added a comment for why that is needed in the first place.